### ISSUE TITLE
CloudFlare R2 (http)

### ISSUE THREAD

**Issue by juretopolak** [unjs/unstorage#318]
### Describe the feature

> This is an experimental driver! This driver only works in a cloudflare worker environment and cannot be used in other runtime environments such as Node.js (r2-http driver is coming soon)

I would just like to ask if you have any approx. date when r2-http driver will be released? I would like to use CF R2 in the Node environment and I'm wondering if it is better to wait for a "native" nuxt/nitro/unstorage driver or the release is not close so it's better to go with another solution? 🙂

Thanks!

### Additional information

- [ ] Would you be willing to help implement this feature?
```[tasklist]
### Tasks
```

**Comment by pi0** [comment]
Hi dear @juretopolak. Feel free to draft a PR if you are willing to help supporting this driver earlier 👍🏼

**Comment by juretopolak** [comment]
Hi @pi0. I'm actually a TypeScript newbie, so I doubt my PR would be useful at all. 🙂 Maybe someday in the future... until then I'll keep learning and admiring the work you experts do. 👌

**Comment by cosbgn** [comment]
Hi @juretopolak  I had to deal with this recently, I found that the best way is to use `aws4fetch` which works on all environments, including workers.
You need something like this, inside for example `/api/upload-file.js`:

```js
import { AwsClient } from 'aws4fetch'
const aws = new AwsClient({ accessKeyId, secretAccessKey })
// Then you can just send a PUT / GET / DELETE request
await aws.fetch(R2_BUCKET_URL, { body: buffer, method: 'PUT', ContentType: type })
```
It's super easy with `aws4fetch` - With native $fetch is more complex because you need to somehow generate the `AWS Signature Version 4`

**Comment by pi0** [comment]
Really nice find. I certainly like to consider aws4fetch. There are small items (such as crypto polyfill to use uncrypto) we can either (temporarily) inline impl to make those or work with @mhart if you would like to make it happen!

**Comment by MickL** [comment]
I guess R2 http is kinda rare because it is an extra overhead if you upload a file to your API (which is not hosted on a Cloudflare worker) and then send the file to Cloudflare by another "upload". If you use Cloudflare R2 it makes sense if your api also runs und Cloudflare so there is no extra transfer from your API to R2.

If you cant deploy to Cloudflare Workers, what you could also do is create an empty Nitro project and deploy it to Cloudflare Pages. Then use this one for file uploads directly from the client or with your api as a proxy.

Btw. since R2 has a S3 api, maybe also this one works?
https://github.com/profilecity/unstorage-s3-driver

**Comment by MickL** [comment]
Wait, according to the docs there is a R2 http driver?
https://unstorage.unjs.io/drivers/cloudflare#cloudflare-kv-http

**Comment by pi0** [comment]
(sorry i mistakenly closed issue. r2 over HTTP driver is not supported yet)

**Comment by pi0** [comment]
S3 supported landed in #361 which is also tested against r2 over HTTP. for now only mentioned in docs in the future we might make a dedicated driver leveraging more native addons of cloudflare r2 if needed.
